refactor DFP cluster to support STRICT_DNS

doujiang24 opened the pull request:
status:
it just works basically, multiple IPs will be chosen randomly in my manual test.
it's a draft PR, suggestions are welcome, thanks!

TODO:
- [x] retire the DNS resolve resolution code.
- [x] honor a TTL configuration, so that the idle cluster could be removed automatically.
- [x] update API proto.
- [x] update & add tests.
- [ ] update doc.

design:
https://docs.google.com/document/d/12S9c1ZPLFVkGNM0xaP2mtNDwNjwxG0eUvg5WhUuNZ98/edit?usp=sharing

For #10823

doujiang24 commented:
Thanks @mattklein123 

The code is basically working now:
1. works well by test manually on my side
2. some integration tests passed in dynamic_forward_proxy/proxy_filter.cc

It would be very appreciated if you could take a quick look, make sure it is in the right direction, at the high level, though there are many TODOs and issues.

I wrote a simple doc for the API proto changes:
Seems we don’t have to introduce `cluster_template`, instead, the sub clusters can inherit config from the parent DFP cluster.
Could you please take a look? We’d better discuss API changes before implementing them. Thanks!
https://docs.google.com/document/d/12S9c1ZPLFVkGNM0xaP2mtNDwNjwxG0eUvg5WhUuNZ98/edit#heading=h.p8q9joxadi54

mattklein123 commented on source/common/upstream/upstream_impl.h:
```diff
@@ -1037,6 +1041,7 @@ class ClusterInfoImpl : public ClusterInfo,
   const uint32_t max_response_headers_count_;
   LoadBalancerType lb_type_;
   const envoy::config::cluster::v3::Cluster::DiscoveryType type_;
+  const envoy::config::cluster::v3::Cluster original_config_;
```
I'm not sure why this was added here but I think this should be specific to DFP and not leak into the common code.

doujiang24 replied:
it's used for sub cluster inherit cluster config from the parent DFP cluster.
I'll try another proper way, thank you.

doujiang24 replied:
Changed to add the `cluster` methods:
https://github.com/envoyproxy/envoy/pull/26420/files#diff-88c4ee0e7897570d35304eca6746d03781c2eacf20a84557b2c1e8df1c115e18R1201-R1210

1. we save the `DFP cluster` into `cluster_info` when initialize the `cluster` in dynamic_forward_proxy/cluster.cc:
https://github.com/envoyproxy/envoy/pull/26420/files#diff-8f0201f1ad64a4d5abf5d3cfe3388d19feea791341f62ebc32064622267da65aR434
2. then, we can get the DFP cluster in worker thread.

mattklein123 reviewed (changes requested):
At a very high level this seems like it's on the right track. A few comments:
1) This is too risky to universally turn on. The new code will need to be runtime guard switched from the old code. This will also make the new code easier to review.
2) If we are going to do the API changes for the template, it would be nice to better understand how that will look vs. all of the hard coding that is happening here. Can you please propose the API changes so we can review those?

Thank you.

/wait

doujiang24 commented:
Thanks, it's really a good idea to introduce a runtime guard.
we don't have to keep strict backward compatibility, at the first step, with the new runtime guard.
it makes the api changes much easier & cleaner.

re the API proposal, here is it with the runtime guard, please see the `API changes` section.
https://docs.google.com/document/d/12S9c1ZPLFVkGNM0xaP2mtNDwNjwxG0eUvg5WhUuNZ98/edit#heading=h.qjireztwkqv6

the biggest change from #10823 is, we don't introduce `cluster_template`, instead, the sub clusters inherit config from the parent DFP cluster.

doujiang24 commented:
I have added a runtime guard `envoy.reloadable_features.enable_strict_dns_sub_cluster_for_dfp_cluster`, and the old code are all keeped.

Just as a reminder, the current PR has not implement the new API proto changes, described in the doc:
https://docs.google.com/document/d/12S9c1ZPLFVkGNM0xaP2mtNDwNjwxG0eUvg5WhUuNZ98/edit#heading=h.qjireztwkqv6

@mattklein123 Could you please take a look at the API changes proposal, thank you very much.

ps: I have squashed & force push, to reduce merge conficts, sorry for the inconvenience.

mattklein123 commented:
We don't need a runtime guard if this is going to be config driven, since the new code should only kick in when there is config. Can you just add in the API changes without the implementation and we can discuss here? Thanks.

/wait

mattklein123 commented on envoy/upstream/upstream.h:
```diff
@@ -1192,6 +1198,16 @@ class ClusterInfo : public Http::FilterChainFactory {
    */
   virtual Http::HeaderValidatorPtr makeHeaderValidator(Http::Protocol protocol) const PURE;
 
+  /**
+   * Setting the cluster that owning this info, only for dynamic forward proxy cluster.
+   */
+  virtual void cluster(ClusterSharedPtr) const PURE;
+
+  /**
+   * @return the cluster that owning this info, only for dynamic forward proxy cluster.
+   */
+  virtual ClusterSharedPtr cluster() const PURE;
```
I still don't understand why you need any interface changes. I don't think any should be required. You are going to be making a new cluster manually given the configuration which you should have access to from the DFP configuration. Can you please remove all of these changes or explain why they are needed?

doujiang24 replied:
It's used to get the `DFP cluster` in `DFP http filer`:
https://github.com/envoyproxy/envoy/pull/26420/files#diff-960724c79345d5bd91d33c3108b549b8c50ca588c7bc57d65f359df841edacfcR272-R277

in the `DFP http filter`, we only have the DFP cluster name, and its `const cluster_info`.
this `cluster` api is used to get the `DFP cluster` from its `const cluster_info`.

why we need the `DFP cluster` in `DFP http filter`?
1. we need the new sub cluster configuration, inherit from the `DFP cluster`. since the `DFP http filter` doesn't have the cluster configurations.
    see: https://github.com/envoyproxy/envoy/pull/26420/files#diff-960724c79345d5bd91d33c3108b549b8c50ca588c7bc57d65f359df841edacfcR65-R66

2. we store the sub cluster name into `cluster_map_` in DFP cluster, so that we can limit the max number of sub clusters, and purge them when they are expired.
    see: https://github.com/envoyproxy/envoy/pull/26420/files#diff-8f0201f1ad64a4d5abf5d3cfe3388d19feea791341f62ebc32064622267da65aR96-R111

Or maybe, is there another proper way to get the `DFP cluster` in `DFP http filter`?
I know the `cluster` API is bit hacky here, but, this is the most proper way I figure out for now.
suggestions are welcome, thank you very much.

doujiang24 replied:
Oh, I see, thanks.

doujiang24 commented:
Thank you! I have added the API changes in this commit:
https://github.com/envoyproxy/envoy/pull/26420/commits/90116dd98f52f63dc02621ce46135f6879df0622

> We don't need a runtime guard if this is going to be config driven, since the new code should only kick in when there is config.

Yep, it does make sense, it's simpler. I'll remove the runtime guard, when we are get on the same page on the new API changes.

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -53,4 +56,21 @@ message ClusterConfig {
   //
   bool allow_coalesced_connections = 3;
+
+  // Enable strict dns sub cluster instead logic dns.
```
Can you flesh this out. From an end user perspective it's hard to understand what this means and why I would use it.

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -53,4 +56,21 @@ message ClusterConfig {
+}
+
+message SubClustersConfig {
+  // The maximum number of sub clusters that the DFP cluster will hold. If not specified defaults to 1024.
+  google.protobuf.UInt32Value max_sub_clusters = 1 [(validate.rules).uint32 = {gt: 0}];
```
Can't this be controlled by the existing circuit breaker configuration? Why do we need a new value here?

mattklein123 commented on api/envoy/extensions/filters/http/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto:
```diff
@@ -26,14 +28,16 @@ message FilterConfig {
   // :repo:`upstream_address.h<source/common/stream_info/upstream_address.h>`).
   bool save_upstream_address = 2;
+
+  // The timeout used for sub cluster initialization. Defaults to 5s if not set.
+  google.protobuf.Duration cluster_init_timeout = 3 [(validate.rules).duration = {gt {}}];
```
Can't this be governed by existing timeouts?

mattklein123 replied:
Ping on this.

doujiang24 replied:
oh, sorry. it's the same reason as the above`max_sub_clusters`.
I think we'd better not let the users to mix use `cluster_init_timeout` and `dns_cache_config`.

`dns_cache_config` is for the old implementation, while `cluster_init_timeout` is for the new implementation.

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -53,4 +56,21 @@ message ClusterConfig {
+  SubClustersConfig sub_clusters_config = 4;
+}
+
+message SubClustersConfig {
```
If we aren't going to support full templates, I would structure this as StrictDnsSubClusterConfig, and put this in a oneof above, so that we can potentially add templates later.

doujiang24 replied:
Added `type` and `lb_policy` for the sub cluster config, they are the necessary template configuration.

mattklein123 reviewed (changes requested):
Thanks left a few comments.

/wait

doujiang24 commented:
Thanks @mattklein123 
All comments are addressed, and already implemented the new API completely. PTAL.
I'll finish the rest test & doc work if the API is ok.

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -23,11 +28,27 @@ message ClusterConfig {
+
+    // Configuration for sub clusters, when this configuration is enabled,
+    // Envoy will create a independent sub cluster dynamically for each host:port.
+    // Most of the configuration of a sub cluster is inherit of the current cluster,
```
```suggestion
    // Most of the configuration of a sub cluster is inherited from the current cluster,
```

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -54,3 +75,26 @@ message ClusterConfig {
+// [#next-free-field: 6]
+message SubClustersConfig {
+  // The :ref:`service discovery type <arch_overview_service_discovery_types>`
+  // to use for resolving a sub cluster. Only STRICT_DNS and LOGICAL_DNS are supported here.
```
What is the point of supported LOGICAL_DNS? Isn't this basically the same as what is already done? Should we just support STRICT_DNS for now?

doujiang24 replied:
yep, only support `STRICT_DNS` for now does make sense.
it could be useful when we deprecate the current API.
will mark this field as `not-implemented-hide`.

mattklein123 reviewed (changes requested):
In general this API makes sense, let's go ahead and add tests and then I can do a full review, thanks.

/wait

doujiang24 commented:
Thank you, I'll add tests in the next few days.

doujiang24 commented:
/retest

doujiang24 commented:
@mattklein123 Tests added, please take another look, thank you. I'll update doc if code is ok.

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -54,3 +75,29 @@ message ClusterConfig {
+
+// [#next-free-field: 6]
+message SubClustersConfig {
+  // The :ref:`service discovery type <arch_overview_service_discovery_types>`
+  // to use for resolving a sub cluster. Hard code STRICT_DNS now.
+  // And we may also support LOGICAL_DNS, when we deprecate the dns_cache_config configuration.
+  //
+  // [#not-implemented-hide:]
+  config.cluster.v3.Cluster.DiscoveryType type = 1 [(validate.rules).enum = {defined_only: true}];
```
Just remove this completely and document that the sub-cluster is hard coded to STRICT_DNS.

mattklein123 reviewed (changes requested):
Some more API comments, thanks.

/wait

doujiang24 commented:
/retest

mattklein123 commented on api/envoy/extensions/filters/http/dynamic_forward_proxy/v3/dynamic_forward_proxy.proto:
```diff
@@ -25,19 +25,21 @@ message FilterConfig {
+    common.dynamic_forward_proxy.v3.DnsCacheConfig dns_cache_config = 1;
+
+    // The timeout used for sub cluster initialization. Defaults to 5s if not set.
+    google.protobuf.Duration cluster_init_timeout = 3 [(validate.rules).duration = {gt {}}];
```
Let's move this into a dedicated SubClusterConfig message so we can add more configuration in the future, thanks.

/wait

yanavlasov commented:
ping @mattklein123 for any more comments

mattklein123 commented on api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto:
```diff
@@ -54,3 +75,23 @@ message ClusterConfig {
 }
+
+// Configuration for sub clusters. Hard code STRICT_DNS cluster type now.
+// We may also support LOGICAL_DNS, when we deprecate the dns_cache_config configuration.
```
Realistically we are never going to deprecate it, so let's just remove this line.

mattklein123 commented on envoy/upstream/upstream.h:
```diff
@@ -1265,8 +1265,38 @@ class Cluster {
 };
 
 using ClusterSharedPtr = std::shared_ptr<Cluster>;
+using ClusterWeakPtr = std::weak_ptr<Cluster>;
 using ClusterConstOptRef = absl::optional<std::reference_wrapper<const Cluster>>;
 
+class DfpCluster {
+public:
+  virtual ~DfpCluster() = default;
+
+  /**
+   * @return the cluster enabled subCluster configuration or not.
+   */
+  virtual bool enableSubCluster() const PURE;
+
+  /**
+   * Create a full cluster config for the cluster_name, with the specified host and port.
+   * @return true and nullptr when the subCluster with the specified cluster_name already
+   *         created, or true and the created cluster config when it not exists and not
+   *         reach the limitation of max_sub_clusters, otherwise, return false and nullptr.
+   */
+  virtual std::pair<bool, absl::optional<envoy::config::cluster::v3::Cluster>>
+  createSubClusterConfig(const std::string& cluster_name, const std::string& host,
+                         const int port) PURE;
+
+  /**
+   * Update the last used time of the subCluster with the specified cluster_name.
+   * @return true if the subCluster is existing.
+   */
+  virtual bool touch(const std::string& cluster_name) PURE;
+};
+
+using DfpClusterSharedPtr = std::shared_ptr<DfpCluster>;
+using DfpClusterWeakPtr = std::weak_ptr<DfpCluster>;
```
There should be no changes to the common headers for this feature. Please remove this and make all of this local to the DFP code where it can be shared via singleton manager or some other mechanism.

mattklein123 commented on source/extensions/clusters/dynamic_forward_proxy/cluster.cc:
```diff
@@ -43,6 +83,132 @@ void Cluster::startPreInit() {
+  // try again to get the sub cluster.
+  auto cluster = cm_.getThreadLocalCluster(cluster_name);
+  if (cluster == nullptr) {
+    ENVOY_LOG(error, "cluster='{}' get thread local failed, too short ttl?", cluster_name);
```
This could log spam, so let's not use an error log here. I'm not quite sure how this can reasonably happen so I would do a debug log and possibly an ENVOY_BUG if you are so inclined?

doujiang24 replied:
I think it might be possible in such a case:
1. we touch the cluster with a short ttl, in the DFP filter, https://github.com/envoyproxy/envoy/blob/0497bd56aef3913ec09daafce3a047b5847fc0ff/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L303-L308
2. the cluster should be there, when first time chooseHost,
3. but it might be removed by the main thead, due to the short TTL,
4. and may fail to get the cluster, the next time chooseHost, when retry upstream request, after a bit long time.

this error log could only happens in a very low rate, and it worth an error level log, if the case is possible.

mattklein123 replied:
I could see this only being possible if the TTL is unreasonably short (should there be some minimum in the config?). Either way, I don't think an error log is a code idea in the data plane path. Either switch to debug, or you could use a backoff warning log.

doujiang24 replied:
it might be several seconds between two retry, may be up to the timeout configuration.
I'm not sure what could be a proper minimum value for the TTL.

Maybe, using a warning log could be good enough?

mattklein123 reviewed (changes requested):
Flushing a few more comments, please also add a release note, thank you.

/wait

doujiang24 commented:
/docs

doujiang24 commented:
Thanks @mattklein123 All addressed, PTAL

mattklein123 commented on envoy/upstream/upstream.h:
```diff
@@ -1265,6 +1265,7 @@ class Cluster {
 };
 
 using ClusterSharedPtr = std::shared_ptr<Cluster>;
+using ClusterWeakPtr = std::weak_ptr<Cluster>;
```
unused?

doujiang24 replied:
oh, sorry, done.

mattklein123 commented on source/extensions/clusters/dynamic_forward_proxy/cluster.h:
```diff
@@ -145,13 +173,29 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
+  mutable absl::Mutex cluster_map_lock_;
+  ClusterInfoMap cluster_map_ ABSL_GUARDED_BY(cluster_map_lock_);
+
+  // only for isShutdown.
```
What does this mean?

doujiang24 replied:
`only for using tls_.isShutdown() in Cluster::~Cluster`, also updated the comment.

mattklein123 commented on source/extensions/common/dynamic_forward_proxy/cluster_store.h:
```diff
@@ -0,0 +1,65 @@
+using DfpClusterSharedPtr = std::shared_ptr<DfpCluster>;
+using DfpClusterWeakPtr = std::weak_ptr<DfpCluster>;
+
+class DFPClusterStore {
+public:
+  // Load the dynamic forward proxy cluster from this store.
+  static DfpClusterSharedPtr load(std::string cluster_name);
+
+  // Save the dynamic forward proxy cluster into this store.
+  static void save(const std::string cluster_name, DfpClusterSharedPtr cluster);
+
+  // Remove the dynamic forward proxy cluster from this store.
+  static void remove(std::string cluster_name);
+
+private:
+  using ClusterMapType = absl::flat_hash_map<std::string, DfpClusterWeakPtr>;
+  struct ClusterStoreType {
+    ClusterMapType map_ ABSL_GUARDED_BY(mutex_);
+    absl::Mutex mutex_;
+  };
+
+  static ClusterStoreType& getClusterStore() { MUTABLE_CONSTRUCT_ON_FIRST_USE(ClusterStoreType); }
+};
```
We shouldn't be using statics for this. Use SingletonManager, which I think may solve your shutdown problem.

doujiang24 replied:
changed to use `SingletonManager`, but still need `tls_.isShutdown` in `Cluster::~Cluster`.

without the `tls_.isShutdown()`, test will hit `assert failure: !shutdown_` while removing cluster.
And here is the full log:
```
[2023-05-08 07:01:15.527][12][info][testing] [test/integration/server.cc:255] stopping integration test server
[2023-05-08 07:01:15.527][233][info][main] [source/server/server.cc:1005] shutting down server instance

... more logs ...

[2023-05-08 07:01:15.529][233][debug][upstream] [source/extensions/clusters/dynamic_forward_proxy/cluster.cc:61] cluster='DFPCluster:sni.lyft.com:443' removing from cluster_map & cluster manager
[2023-05-08 07:01:15.530][233][debug][init] [source/common/init/watcher_impl.cc:31] ClusterImplBase destroyed
[2023-05-08 07:01:15.530][233][debug][init] [source/common/init/watcher_impl.cc:31] init manager Cluster DFPCluster:sni.lyft.com:443 destroyed
[2023-05-08 07:01:15.530][233][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:808] removing cluster DFPCluster:sni.lyft.com:443
[2023-05-08 07:01:15.530][233][critical][assert] [source/common/thread_local/thread_local_impl.cc:152] assert failure: !shutdown_.
```

maybe using `tls_.isShutdown` is acceptable here?

mattklein123 reviewed (changes requested):
/wait

RyanTheOptimist commented:
@mattklein123 friendly ping

mattklein123 commented on source/extensions/clusters/dynamic_forward_proxy/cluster.cc:
```diff
@@ -173,7 +173,7 @@ Upstream::HostConstSharedPtr Cluster::chooseHost(absl::string_view host,
   auto cluster = cm_.getThreadLocalCluster(cluster_name);
   if (cluster == nullptr) {
-    ENVOY_LOG(error, "cluster='{}' get thread local failed, too short ttl?", cluster_name);
+    ENVOY_LOG(warn, "cluster='{}' get thread local failed, too short ttl?", cluster_name);
```
This can still log spam. Please use the backoff logger if you want to keep as warning.

mattklein123 commented on source/extensions/clusters/dynamic_forward_proxy/cluster.h:
```diff
@@ -145,13 +173,29 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
+  mutable absl::Mutex cluster_map_lock_;
+  ClusterInfoMap cluster_map_ ABSL_GUARDED_BY(cluster_map_lock_);
+
+  // only for using tls_.isShutdown() in Cluster::~Cluster().
+  ThreadLocal::TypedSlot<ThreadLocalConfig> tls_;
```
This is kind of strange if only used for the shutdown case. Since you are passed the cluster factory context, can we just expose a method on the server about whether it's shutdown or not (there might already be one) or somehow make the cluster manager just ignore whatever is happening during shutdown to avoid having that leak into this code? Thank you.

mattklein123 reviewed (changes requested):
/wait

doujiang24 commented:
/retest

mattklein123 reviewed (dismissed):
LGTM to ship and iterate if needed. Thank you for the hard work!

mattklein123 commented:
Oops sorry looks like this needs a main merge.

/wait

doujiang24 commented:
/retest

doujiang24 commented:
Great, thank you very much! @mattklein123
